File: Napakalaki/src/NapakalakiGame/SpecificBadConsequence.java
/**
 * SpecificBadConsequence.java
 */

package NapakalakiGame;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Arrays;

/**
 * Representa un mal rollo de tesoros específicos.
 */
public class SpecificBadConsequence extends BadConsequence {
    
    private ArrayList<TreasureKind> specificHiddenTreasures;
    private ArrayList<TreasureKind> specificVisibleTreasures;
    
    public SpecificBadConsequence(String t, int l, ArrayList<TreasureKind> v,
                                  ArrayList<TreasureKind> h) {
        super(t, l);
        specificVisibleTreasures = v;
        specificHiddenTreasures = h;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }
    
    @Override
    public boolean isEmpty() {
        return specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty();
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (!specificVisibleTreasures.isEmpty()) {
            specificVisibleTreasures.remove(t.getType());
        }
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (!specificHiddenTreasures.isEmpty()) {
            specificHiddenTreasures.remove(t.getType());
        }
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, 
                                                   ArrayList<Treasure> h) {
        // Nuevos arrays con los tipos de tesoros
        ArrayList<TreasureKind> newVisibleType = new ArrayList();
        ArrayList<TreasureKind> newHiddenType = new ArrayList();
            
        for (Treasure t : v) {
            newVisibleType.add(t.getType());
        }
            
        for (Treasure t : h) {
            newHiddenType.add(t.getType());
        }
            
        // Nuevos arrays con la intersección (con duplicados)
        ArrayList<TreasureKind> newVisible = new ArrayList();
        ArrayList<TreasureKind> newHidden = new ArrayList();
            
        for (TreasureKind tKind : TreasureKind.values()) {
            int min_visible = Math.min(Collections.frequency(specificVisibleTreasures, tKind),
                                       Collections.frequency(newVisibleType, tKind));
            int min_hidden = Math.min(Collections.frequency(specificHiddenTreasures, tKind),
                                      Collections.frequency(newHiddenType, tKind));
                
            newVisible.addAll(Collections.nCopies(min_visible, tKind));
            newHidden.addAll(Collections.nCopies(min_hidden, tKind));
        }
            
        return new SpecificBadConsequence(getText(), getLevels(), newVisible, newHidden);
    }
    
    @Override
    public String toString() {
        return super.toString() + "\n    * Tesoros visibles específicos perdidos: " + 
               Arrays.toString(specificVisibleTreasures.toArray()) +
               "\n    * Tesoros ocultos específicos perdidos: " + 
               Arrays.toString(specificHiddenTreasures.toArray());
    }
    
}

File: Napakalaki/lib/specific_bad_consequence.rb
# encoding: utf-8

require_relative 'treasure_kind'
require_relative 'bad_consequence'

module NapakalakiGame

  # class SpecificBadConsequence: representa un mal rollo de tesoros específicos
  class SpecificBadConsequence < BadConsequence
    def initialize(t, l, v, h)
      super(t, l)
      @specificVisibleTreasures = v
      @specificHiddenTreasures = h
    end

    def getSpecificVisibleTreasures
      @specificVisibleTreasures
    end

    def getNHiddenTreasures
      @specificHiddenTreasures
    end

    def isEmpty
      @specificVisibleTreasures == [] and @specificHiddenTreasures == []
    end

    def substractVisibleTreasure(t)
      type = t.type
      if (not @specificVisibleTreasures.empty?) \
            and @specificVisibleTreasures.include?(type)
        @specificVisibleTreasures.delete_at(@specificVisibleTreasures.index(type))
      end
    end
    
    def substractHiddenTreasure(t)
      type = t.type
      if (not @specificHiddenTreasures.empty?) \
            and @specificHiddenTreasures.include?(type)
        @specificHiddenTreasures.delete_at(@specificHiddenTreasures.index(type))
      end
    end

    def adjustToFitTreasureLists(v,h)
      # El array que nos pasan es de tesoros pero lo necesitamos de TIPOS de tesoros
      vTypes = []
      v.each do |t|
        vTypes << t.type
      end

      newSpecificVisibleTreasures = @specificVisibleTreasures & vTypes
      if @specificVisibleTreasures.count(TreasureKind::ONEHAND) == 2 and \
          vTypes.count(TreasureKind::ONEHAND) >= 2
        newSpecificVisibleTreasures << TreasureKind::ONEHAND
      end

      # Igual que antes, necesitamos los TIPOS de los tesoros
      hTypes = []
      h.each do |t|
        hTypes << t.type
      end

      # Intersección, manteniendo duplicados (flat_map concatena)
      newSpecificHiddenTreasures = (@specificHiddenTreasures & hTypes).flat_map { |n| [n]*[@specificHiddenTreasures.count(n), hTypes.count(n)].min }

      SpecificBadConsequence.new(@text, @levels, newSpecificVisibleTreasures, newSpecificHiddenTreasures)
    end

    def to_s
      text = super
      text << "\n    * Tesoros específicos visibles perdidos: \
              #{@specificVisibleTreasures}"
      text << "\n    * Tesoros específicos ocultos perdidos: \
              #{@specificHiddenTreasures}"
    end

  end
  
end